lib/cloudera-manager/service.rb

```diff
@@ -1,4 +1,64 @@
 module ClouderaManager
   class Service < BaseResource
+    class RestartFailedException < BaseException; end
+
+    belongs_to :cluster
+    has_many :roles
+    primary_key :name
+    collection_path "clusters/:cluster_id/services"
+    after_find { |s| s.cluster_id = s.clusterRef[:clusterName] }
+
+    BULK_RESTART_SAFE_ROLE_TYPES = [
+      "IMPALAD",
+      "DATANODE",
+      "NODEMANAGER"
+    ]
+
+    RESTART_ORDER = {
+      "YARN" => ["JOBHISTORY", "NODEMANAGER", "RESOURCEMANAGER"],
+      "HDFS" => ["SECONDARYNAMENODE", "HTTPFS", "FAILOVERCONTROLLER", "NFSGATEWAY", "DATANODE", "NAMENODE", "JOURNALNODE"],
+      "IMPALA" => ["IMPALAD", "STATESTORE", "CATALOGSERVER", "LLAMA"]
+    }
+
+    def enterprise_rolling_restart!
+      remote_command!("/commands/rollingRestart")
+    end
+
+    def rolling_restart!(bulk_batch_size)
+      ordered_role_types = RESTART_ORDER[self.type] || self.role_types
+      ordered_role_types.each_with_object(ClouderaManager::BulkCommand.new) do |role_type, all_commands|
+        batch_size = BULK_RESTART_SAFE_ROLE_TYPES.include?(role_type) ? bulk_batch_size : 1
+        all_commands.concat rolling_restart_of_type!(role_type, batch_size)
+      end
+    end
+
+    def rolling_restart_of_type!(role_type, batch_size)
+      ClouderaManager.logger.info "Starting rolling restart of #{self.name} - #{role_type}"
+
+      all_commands = ClouderaManager::BulkCommand.new
+      restartable_role_names_for_type(role_type).each_slice(batch_size) do |slice|
+        ClouderaManager.logger.info "Restarting #{slice.join(', ')}"
+
+        commands = remote_command!("/roleCommands/restart", items: slice)
+        if commands.failed.size > 0
+          ClouderaManager.logger.warn "Restart of #{commands.failed.map(&:name).join(', ')} all failed!"
+        end
+        if commands.success_ratio < 0.1
```

Comment:
Word ya. The batch size for any of the masters I want to always be 1, which means if it failed to come back the success ratio will be 0. That said I think an active health check is in order for those roles instead of just verifying that the command succeeded since I think that just means the process started.